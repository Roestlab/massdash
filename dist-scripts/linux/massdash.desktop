[Desktop Entry]
Version=0.0.8
Type=Application
Name=massdash
Exec=/usr/local/bin/massdash
Icon=/usr/share/icons/massdash.ico
Terminal=true
