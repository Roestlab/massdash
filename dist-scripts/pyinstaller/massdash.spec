# -*- mode: python ; coding: utf-8 -*-

import site
import importlib.metadata
import os
import sys

from PyInstaller.building.build_main import Analysis, PYZ, EXE, COLLECT
from PyInstaller.utils.hooks import collect_all, collect_data_files, copy_metadata

from transformers.dependency_versions_check import pkgs_to_check_at_runtime

sys.setrecursionlimit(sys.getrecursionlimit() * 5)

##################### User definitions
exe_name = 'massdash_gui'
script_name = 'run.py'
if sys.platform[:6] == "darwin":
	icon = '../../massdash/assets/img/MassDash_Logo.icns'
else:
	icon = '../../massdash/assets/img/MassDash_Logo.ico'
block_cipher = None
location = os.getcwd()
project = "massdash"
remove_tests = True
bundle_name = "massdash"
#####################


requirements = {project, "streamlit_javascript", "upsetplot", "distributed"}
datas = [(f"{site.getsitepackages()[0]}/streamlit/runtime", "./streamlit/runtime")]
hidden_imports = set()
binaries = []
checked = set()
while requirements:
	requirement = requirements.pop()
	print(f"Info: Checking {requirement}")
	checked.add(requirement)
	if requirement in ["pywin32"]:
		continue
	try:
		module_version = importlib.metadata.version(requirement)
	except (
		importlib.metadata.PackageNotFoundError,
		ModuleNotFoundError,
		ImportError
	):
		continue
	try:
		datas_, binaries_, hidden_imports_ = collect_all(
			requirement,
			include_py_files=True
		)
	except ImportError:
		continue
	datas += datas_
	# binaries += binaries_
	hidden_imports_ = set(hidden_imports_)
	if "" in hidden_imports_:
		hidden_imports_.remove("")
	if None in hidden_imports_:
		hidden_imports_.remove(None)
	requirements |= hidden_imports_ - checked
	hidden_imports |= hidden_imports_

if remove_tests:
	hidden_imports = sorted(
		[h for h in hidden_imports if "test" not in h.split(".")]
	)
else:
	hidden_imports = sorted(hidden_imports)


hidden_imports = [h for h in hidden_imports if "__pycache__" not in h]
datas = [d for d in datas if ("__pycache__" not in d[0]) and (d[1] not in [".", "Resources", "scripts"])]

#if sys.platform[:5] == "win32":
#	base_path = os.path.dirname(sys.executable)
#	library_path = os.path.join(base_path, "Library", "bin")
#	dll_path = os.path.join(base_path, "DLLs")
#	libcrypto_dll_path = os.path.join(dll_path, "libcrypto-1_1-x64.dll")
#	libssl_dll_path = os.path.join(dll_path, "libssl-1_1-x64.dll")
#	libcrypto_lib_path = os.path.join(library_path, "libcrypto-1_1-x64.dll")
#	libssl_lib_path = os.path.join(library_path, "libssl-1_1-x64.dll")
#	if not os.path.exists(libcrypto_dll_path):
#		datas.append((libcrypto_lib_path, "."))
#	if not os.path.exists(libssl_dll_path):
#		datas.append((libssl_lib_path, "."))

for _pkg in ["python","accelerate"]:
	if _pkg in pkgs_to_check_at_runtime:
		pkgs_to_check_at_runtime.remove(_pkg)
for _pkg in pkgs_to_check_at_runtime:
	datas += copy_metadata(_pkg)


datas += collect_data_files("streamlit")
datas += copy_metadata("streamlit")
datas += collect_data_files("pyopenms")
datas += copy_metadata("pyopenms")
datas += collect_data_files("massdash")
datas += copy_metadata("massdash")

hidden_imports = ['pyopenms', 'massdash', 'torchaudio.lib.libtorchaudio']

a = Analysis(
	[script_name],
	pathex=[location],
	binaries=binaries,
	datas=datas,
	hiddenimports=hidden_imports,
	hookspath=['./hooks'],
	runtime_hooks=[],
	excludes=[],
	win_no_prefer_redirects=False,
	win_private_assemblies=False,
	cipher=block_cipher,
	noarchive=False
)
pyz = PYZ(
	a.pure,
	a.zipped_data,
	cipher=block_cipher
)

if sys.platform[:5] == "linux":
	exe = EXE(
		pyz,
		a.scripts,
		a.binaries,
		a.zipfiles,
		a.datas,
		name=bundle_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		upx_exclude=[],
		icon=icon
	)
else:
	exe = EXE(
		pyz,
		a.scripts,
		# a.binaries,
		a.zipfiles,
		# a.datas,
		exclude_binaries=True,
		name=exe_name,
		debug=False,
		bootloader_ignore_signals=False,
		strip=False,
		upx=True,
		console=True,
		icon=icon
	)
	coll = COLLECT(
		exe,
		a.binaries,
		# a.zipfiles,
		a.datas,
		strip=False,
		upx=True,
		upx_exclude=[],
		name=exe_name
	)