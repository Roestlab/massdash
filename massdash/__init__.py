"""
massdash
~~~~~~~~
TODO: ADD general documentation about how to use the auto API here

Package Structure
-----------------
testing - a test package
"""

__all__ = [ 'structs', 'preprocess', 'loaders', 'ui', 'server', 'testing' ]
__version__ = "0.0.9"
