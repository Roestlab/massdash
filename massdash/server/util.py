"""
massdash/server/util
~~~~~~~~~~~~~~~~~~~~
"""

import mmap

def get_string_mslevels_from_bool(mslevel_bool_dict):
    """
    Converts a dictionary of boolean values for mslevels to a string.

    Args:
    mslevel_bool_dict : dict
        A dictionary containing boolean values for mslevels.

    Returns:
    mslevel_str : str
        A string representing the selected mslevel.
    """
    if mslevel_bool_dict['ms1']:
        mslevel_str = 'ms1'
    elif mslevel_bool_dict['ms2']:
        mslevel_str = 'ms2'
    elif mslevel_bool_dict['ms1'] and mslevel_bool_dict['ms2']:
        mslevel_str = 'ms1ms2'
    else:
        raise ValueError('No mslevel selected')
    return mslevel_str

def check_ion_mobility(mzml_file, num_lines_to_check=10_000_000):
    """
    Check if the mzML file contains ion mobility data

    Args:
    mzml_file: (str) mzML file to load
    num_lines_to_check: (int) Number of lines to check for "Ion Mobility"

    Returns:
    Return a boolean indicating if the mzML file contains ion mobility data
    """
    has_ion_mobility = False
    with open(mzml_file, 'rb', 0) as file:
        s = mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) 
        for line in iter(s.readline, b""):
            if b'Ion Mobility' in line:
                has_ion_mobility = True
                break 
            if s.tell() > num_lines_to_check:
                break
    return has_ion_mobility