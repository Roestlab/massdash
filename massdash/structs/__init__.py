"""
massdash/structs
~~~~~~~~~~~~~~~~

Internal Data Structures for storing data 
"""

from .Chromatogram import Chromatogram
from .Data1D import Data1D
from .FeatureMap import FeatureMap
from .FeatureMapCollection import FeatureMapCollection
from .GenericFeature import GenericFeature
from .GenericStructCollection import GenericStructCollection
from .Mobilogram import Mobilogram
from .Spectrum import Spectrum
from .TargetedDIAConfig import TargetedDIAConfig
from .TopTransitionGroupFeatureCollection import TopTransitionGroupFeatureCollection
from .TransitionFeature import TransitionFeature
from .TransitionGroupFeature import TransitionGroupFeature
from .TransitionGroup import TransitionGroup
from .TransitionGroupCollection import TransitionGroupCollection
from .TransitionGroupFeatureCollection import TransitionGroupFeatureCollection


__all__ = [ 
            "Chromatogram", 
            "Data1D",
            "FeatureMap", 
            "FeatureMapCollection",
            "GenericFeature",
            "GenericStructCollection",
            "Mobilogram", 
            "TargetedDIAConfig",
            "TransitionFeature", 
            "TransitionGroup", 
            "TopTransitionGroupFeatureCollection",
            "TransitionGroupCollection",
            "TransitionGroupFeature", 
            "TransitionGroupFeatureCollection",
            "Spectrum"]
