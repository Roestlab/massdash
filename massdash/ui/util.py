"""
massdash/ui/util
~~~~~~~~~~~~~~~~
"""

from os import getcwd
from os.path import dirname
import streamlit as st
from tkinter import Tk, filedialog

def clicked(button):
    """
    Updates the session state to indicate that a button has been clicked.

    Args:
        button (str): The name of the button that was clicked.

    Returns:
        None
    """
    st.session_state.clicked[button] = True
    
def st_mutable_write(text=None):
    """
    Creates a mutable container that can be updated with text.
    
    Args:
        text (str): The text to be written to the container.
        
    Returns:
        st.empty(): The mutable container.
    """
    write_container = st.empty()
    if text is not None:
        write_container.write(text)
    return write_container

def get_parent_directory(input_file: str=None):
    """
    Get the parent directory of a file, if no file is provided, the current working directory is returned.

    Args:
        input_file (str): The path to the feature file.

    Returns:
        str: The parent directory.
    """
    if input_file is None:
        return getcwd()
    return dirname(input_file)

def tk_file_dialog(file_type: list = [], title: str = "Select File", parent_dir: str = getcwd()):
    """
    Creates a Tkinter file dialog for selecting a file.

    Args:
        file_type (list): List of tuples specifying file types and their corresponding extensions.
                           Example: [("Text files", "*.txt"), ("CSV files", "*.csv")]
        title (str): The title of the file dialog.
        parent_dir (str): The path to the parent directory of the file dialog.

    Returns:
        str: The path to the selected file.
    """
    root = Tk()
    root.withdraw()
    file_type.extend([("All files", "*.*")])
    file_path = filedialog.askopenfilename(filetypes=file_type, title=title, initialdir=parent_dir)
    root.destroy()
    return file_path


def display_input_section(title, key_base: str, file_extension: str, dialog_title: str, placeholder: str, st_cols = None):
    """
    Display an input section with a title, file browsing button, and text input field.

    Parameters:
        title (str): The title of the input section.
        key_base (str): The base key used for storing session state variables
        file_extension (str): The file extension to filter when browsing for files.
        dialog_title (str): The title of the file browsing dialog.
        placeholder (str): The placeholder text for the text input field.
        st_cols (streamlit.columns): The columns to display the input section in.

    Returns:
        str: The value entered in the text input field.
    """
    st.subheader(title)
    if st_cols is None:
        st_cols = st.columns([0.05, 0.95], gap="small")
    with st_cols[0]:
        st.write("\n")
        st.write("\n")
        dialog_button = st.button("üìÅ", key=f'{key_base}_browse', help=f"Browse for the {title} file.")
        if dialog_button:
            st.session_state.tmp_input_dict[key_base] = tk_file_dialog(file_extension, dialog_title, get_parent_directory(st.session_state.tmp_input_dict[key_base]))    
    with st_cols[1]:
        input_value = st.text_input("Enter file path", value=st.session_state.tmp_input_dict[key_base],
                                    placeholder=placeholder, key=f'{key_base}_tmp',
                                    help=f"Path to the {title} file ({file_extension})")
    return input_value
