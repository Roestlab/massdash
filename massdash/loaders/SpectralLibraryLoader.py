"""
massdash/loaders/SpectralLibraryLoader
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

import os
from typing import List
import pandas as pd

# Loaders
from .access.TransitionTSVDataAccess import TransitionTSVDataAccess
from .access.TransitionPQPDataAccess import TransitionPQPDataAccess
# Structs
from ..structs.TransitionGroupFeature import TransitionGroupFeature
# Utils
from ..util import LOGGER

class SpectralLibraryLoader:
    """
    Class for loading spectral library files.
    
    Attributes:
        in_file (str): The path to the spectral library file.
        data (pd.DataFrame): The spectral library data.
        
    Methods:
        load: Loads the spectral library file.
        save: Saves the spectral library data to a file.
        get_unique_proteins: Retrieves a list of unique protein IDs from the spectral library.
        get_unique_peptides_per_protein: Retrieves a list of unique peptide sequences for a given protein.
        get_unique_charge_states_per_peptide: Retrieves a list of unique charge states for a given peptide.
        get_peptide_precursor_mz: Retrieves the precursor m/z value for a given peptide and charge.
        get_peptide_product_mz_list: Retrieves a list of product m/z values for a given peptide and charge.
        get_peptide_product_charge_list: Retrieves a list of product charges for a given peptide and charge.
        get_peptide_retention_time: Retrieves the retention time for a given peptide and charge.
        get_peptide_ion_mobility: Retrieves the ion mobility of a peptide with a specific charge from the spectral library.
        get_peptide_library_intensity: Retrieves the library intensity for a specific peptide and charge.
        get_peptide_fragment_annotation_list: Retrieves a list of fragment annotations for a given peptide and charge.
        filter_for_target_transition_list: Filters the data for a specific target transition list based on the given protein, peptide, and charge.
    """
    def __init__(self, in_file: str, verbose: bool=False) -> None:
        self.in_file = in_file
        self.data: pd.DataFrame = self.load()
        self.has_im = 'PrecursorIonMobility' in self.data.columns and self.data['PrecursorIonMobility'].notnull().any()
        
        LOGGER.name = "SpectralLibraryLoader"
        if verbose:
            LOGGER.setLevel("DEBUG")
        else:
            LOGGER.setLevel("INFO")

    def load(self) -> pd.DataFrame:
        """
        Load the transition list file
        """
        LOGGER.debug(f"Loading transition file {self.in_file}")
        _, file_extension = os.path.splitext(self.in_file)
        if file_extension.lower() == '.tsv':
            return TransitionTSVDataAccess(self.in_file).load()
        elif file_extension.lower() == '.pqp' or file_extension.lower() == '.osw':
            return TransitionPQPDataAccess(self.in_file).load()
        else:
            raise ValueError("Unsupported file format")

    def save(self, file_path: str) -> None:
        """
        Save the data to a file.

        Args:
            file_path (str): The path to the file where the data will be saved.

        Returns:
            None
        """
        self.data.to_csv(file_path, sep='\t', index=False)

    def get_unique_proteins(self) -> List[str]:
        """
        Returns a list of unique protein IDs from the data.

        Returns:
            List[str]: A list of unique protein IDs.
        """
        return self.data['ProteinId'].unique()

    def get_unique_peptides_per_protein(self, protein: str) -> List[str]:
        """
        Returns a list of unique peptide sequences for a given protein.

        Args:
            protein (str): The protein ID.

        Returns:
            List[str]: A list of unique peptide sequences.
        """
        return self.data[self.data['ProteinId'] == protein]['ModifiedPeptideSequence'].unique()

    def get_unique_charge_states_per_peptide(self, peptide: str) -> List[int]:
        """
        Returns a list of unique charge states for a given peptide.

        Args:
            peptide (str): The peptide sequence.

        Returns:
            List[int]: A list of unique charge states associated with the peptide.
        """
        return self.data[self.data['ModifiedPeptideSequence'] == peptide]['PrecursorCharge'].unique()

    def get_peptide_precursor_mz(self, peptide: str, charge: int) -> float:
        """
        Get the precursor m/z value for a given peptide and charge.

        Parameters:
        peptide (str): The modified peptide sequence.
        charge (int): The precursor charge.

        Returns:
        float: The precursor m/z value.
        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['PrecursorMz'].iloc[0]
    
    def get_peptide_product_mz_list(self, peptide: str, charge: int) -> List[float]:
        """
        Retrieve a list of product m/z values for a given peptide and charge.

        Args:
            peptide (str): The modified peptide sequence.
            charge (int): The precursor charge.

        Returns:
            List[float]: A list of product m/z values.

        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['ProductMz'].tolist()
    
    # TODO not used remove?
    def populateTransitionGroupFeature(self, transition_group_feature: TransitionGroupFeature) -> TransitionGroupFeature:
        """
        Appends library information to a TransitionGroupFeature object

        Args:
            transition_group_feature (TransitionGroupFeature): The TransitionGroupFeature object to append library information to.

        Returns:
            TransitionGroupFeature: The TransitionGroupFeature object with appended library information.
        """
        library_data = self.data[(self.data['ModifiedPeptideSequence'] == transition_group_feature.sequence) & (self.data['PrecursorCharge'] == transition_group_feature.precursor_charge)]

        if library_data.empty:
            LOGGER.warning(f"No library data found for {transition_group_feature.sequence} {transition_group_feature.precursor_charge}")
            return transition_group_feature
        transition_group_feature.product_annotations = library_data['Annotation'].tolist()
        transition_group_feature.product_mz = library_data['ProductMz'].tolist()
        transition_group_feature.precursor_mz = library_data['PrecursorMz'].iloc[0]
        return transition_group_feature
    
    def populateTransitionGroupFeatures(self, transition_group_features: List[TransitionGroupFeature]) -> List[TransitionGroupFeature]:
        """
        Populates library information for a list of TransitionGroupFeature objects.
        ***Important:*** Assumes that all TransitionGroupFeature objects have the same peptide sequence and precursor charge.
        """
        peptide_sequence = transition_group_features[0].sequence
        precursor_charge = transition_group_features[0].precursor_charge
        library_data = self.data[(self.data['ModifiedPeptideSequence'] == peptide_sequence) & (self.data['PrecursorCharge'] == precursor_charge)]

        for t in transition_group_features:
            t.product_annotations = library_data['Annotation'].tolist()
            t.product_mz = library_data['ProductMz'].tolist()
            t.precursor_mz = library_data['PrecursorMz'].iloc[0]

    def get_peptide_product_charge_list(self, peptide: str, charge: int) -> List[int]:
        """
        Retrieve a list of product charges for a given peptide and charge.

        Args:
            peptide (str): The modified peptide sequence.
            charge (int): The precursor charge.

        Returns:
            List[int]: A list of product charges.

        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['ProductCharge'].tolist()
    
    def get_peptide_retention_time(self, peptide: str, charge: int) -> float:
        """
        Get the retention time for a given peptide and charge.

        Parameters:
        peptide (str): The modified peptide sequence.
        charge (int): The precursor charge.

        Returns:
        float: The normalized retention time.
        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['NormalizedRetentionTime'].iloc[0]
    
    def get_peptide_ion_mobility(self, peptide: str, charge: int) -> float:
        """
        Retrieves the ion mobility of a peptide with a specific charge from the spectral library.

        Parameters:
        peptide (str): The modified peptide sequence.
        charge (int): The precursor charge.

        Returns:
        float: The precursor ion mobility.
        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['PrecursorIonMobility'].iloc[0]

    def get_peptide_library_intensity(self, peptide: str, charge: int) -> float:
        """
        Get the library intensity for a specific peptide and charge.

        Parameters:
            peptide (str): The modified peptide sequence.
            charge (int): The precursor charge.

        Returns:
            float: The library intensity for the specified peptide and charge.
        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['LibraryIntensity'].iloc[0]

    def get_peptide_fragment_annotation_list(self, peptide: str, charge: int) -> List[str]:
        """
        Retrieves a list of fragment annotations for a given peptide and charge.

        Args:
            peptide (str): The peptide sequence.
            charge (int): The precursor charge.

        Returns:
            List[str]: A list of fragment annotations.
        """
        return self.data[(self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]['Annotation'].tolist()
    
    def filter_for_target_transition_list(self, protein: str, peptide: str, charge: int) -> pd.DataFrame:
        """
        Filters the data for a specific target transition list based on the given protein, peptide, and charge.

        Args:
            protein (str): The protein ID.
            peptide (str): The modified peptide sequence.
            charge (int): The precursor charge.

        Returns:
            pd.DataFrame: The filtered data containing the target transition list.
        """
        return self.data[(self.data['ProteinId'] == protein) & (self.data['ModifiedPeptideSequence'] == peptide) & (self.data['PrecursorCharge'] == charge)]
