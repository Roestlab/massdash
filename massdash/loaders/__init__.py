"""
massdash/loaders
~~~~~~~~~~~~~~~~

The :mod:`massdash.loaders` subpackage contains the structures for loading data into MassDash
"""

from .GenericChromatogramLoader import GenericChromatogramLoader
from .GenericLoader import GenericLoader
from .GenericSpectrumLoader import GenericSpectrumLoader
from .MzMLDataLoader import MzMLDataLoader
from .SpectralLibraryLoader import SpectralLibraryLoader
from .SqMassLoader import SqMassLoader

__all__ = [ 
            "GenericChromatogramLoader",
            "GenericLoader",
            "GenericSpectrumLoader",
            "MzMLDataLoader", 
            "SpectralLibraryLoader",
            "SqMassLoader"]
