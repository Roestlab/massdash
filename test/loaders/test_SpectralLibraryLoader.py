"""
test/loaders/test_SpectralLibraryLoader
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
"""

from pathlib import Path
import pytest

from syrupy.extensions.amber import AmberDataSerializer
from massdash.structs import TransitionGroupFeature
from massdash.loaders.SpectralLibraryLoader import SpectralLibraryLoader
from massdash.util import find_git_directory
from massdash.testing.PandasSnapshotExtension import PandasSnapshotExtenstion

TEST_PATH = find_git_directory(Path(__file__).resolve()).parent / 'test'

@pytest.fixture
def snapshot_pandas(snapshot):
    return snapshot.use_extension(PandasSnapshotExtenstion)


@pytest.fixture
def spectral_library_loader(request):
    if request.param == 'diann':
        return SpectralLibraryLoader(f"{TEST_PATH}/test_data/example_dia/diann/lib/test_1_lib.tsv")
    if request.param == 'openswath':
        return SpectralLibraryLoader(f"{TEST_PATH}/test_data/example_dia/openswath/lib/test.pqp")
    if request.param == 'diann-im':
        return SpectralLibraryLoader(f'{TEST_PATH}/test_data/library/ionMobilityTestLibrary.tsv')

@pytest.fixture
def transition_group_features():
    return [TransitionGroupFeature(sequence="DYASIDAAPEER", precursor_charge=2, leftBoundary=10, rightBoundary=20)] # note left and right boundary not used so just dummy values

@pytest.mark.parametrize('spectral_library_loader', ['diann', 'openswath'], indirect=['spectral_library_loader'])
def test_load(spectral_library_loader, snapshot_pandas):
    data = spectral_library_loader.load()
    assert snapshot_pandas == data

def test_save():
    pass # not testable


@pytest.mark.parametrize('spectral_library_loader', ['diann', 'openswath'], indirect=['spectral_library_loader'])
def test_get_unique_proteins(spectral_library_loader, snapshot):
    proteins = spectral_library_loader.get_unique_proteins()
    assert snapshot == proteins


@pytest.mark.parametrize('spectral_library_loader', ['diann', 'openswath'], indirect=['spectral_library_loader'])
def test_get_unique_peptides_per_protein(spectral_library_loader, snapshot):
    peptides = spectral_library_loader.get_unique_peptides_per_protein("protein_name")
    assert snapshot == peptides


@pytest.mark.parametrize('spectral_library_loader', ['diann', 'openswath'], indirect=['spectral_library_loader'])
def test_get_unique_charge_states_per_peptide(spectral_library_loader, snapshot):
    charge_states = spectral_library_loader.get_unique_charge_states_per_peptide("peptide_sequence")
    assert snapshot == charge_states


@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_precursor_mz(spectral_library_loader, pep, charge, snapshot):
    precursor_mz = spectral_library_loader.get_peptide_precursor_mz(pep, charge)
    assert snapshot == precursor_mz


@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_product_mz_list(spectral_library_loader, pep, charge, snapshot):
    product_mz_list = spectral_library_loader.get_peptide_product_mz_list(pep, charge)
    assert snapshot == product_mz_list

def test_populateTransitionGroupFeature():
    pass # not used so not tested

@pytest.mark.parametrize('spectral_library_loader', ['diann', 'openswath'], indirect=['spectral_library_loader'])
def test_populateTransitionGroupFeatures(spectral_library_loader, transition_group_features, snapshot):
    features = spectral_library_loader.populateTransitionGroupFeatures(transition_group_features)
    assert snapshot == AmberDataSerializer.object_as_named_tuple(features)

@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_product_charge_list(spectral_library_loader, pep, charge, snapshot):
    product_charge_list = spectral_library_loader.get_peptide_product_charge_list(pep, charge)
    assert snapshot == product_charge_list


# note: does not work with DIA-NN since no NormalizedRetentionTime column
@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_retention_time(spectral_library_loader, pep, charge, snapshot):
    retention_time = spectral_library_loader.get_peptide_retention_time(pep, charge)
    assert snapshot == retention_time


@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann-im','AFVDFLSDEIK', 2), ], indirect=['spectral_library_loader'])
def test_get_peptide_ion_mobility(spectral_library_loader, pep, charge, snapshot):
    ion_mobility = spectral_library_loader.get_peptide_ion_mobility(pep, charge)
    assert snapshot == ion_mobility


@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_library_intensity(spectral_library_loader, pep, charge, snapshot):
    library_intensity = spectral_library_loader.get_peptide_library_intensity(pep, charge)
    assert snapshot == library_intensity


@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def test_get_peptide_fragment_annotation_list(spectral_library_loader, pep, charge, snapshot):
    fragment_annotation_list = spectral_library_loader.get_peptide_fragment_annotation_list(pep, charge)
    assert snapshot == fragment_annotation_list


@pytest.mark.parametrize('spectral_library_loader,pep,charge,fragment', [('diann','DYASIDAAPEER', 2, 'b3^1'), ('openswath', 'DYASIDAAPEER', 2, 'y4/0.01y10^2/0.50')], indirect=['spectral_library_loader'])
def test_get_fragment_library_intensity(spectral_library_loader, pep, charge, fragment, snapshot):
    library_intensity = spectral_library_loader.get_fragment_library_intensity(pep, charge, fragment)
    assert snapshot == library_intensity

@pytest.mark.parametrize('spectral_library_loader,pep,charge', [('diann','DYASIDAAPEER', 2), ('openswath', 'DYASIDAAPEER', 2)], indirect=['spectral_library_loader'])
def filter_for_target_transition_list(spectral_library_loader, pep, charge, target_transition_list, snapshot_pandas):
    filtered_transition_list = spectral_library_loader.filter_for_target_transition_list(pep, charge, target_transition_list)
    assert snapshot_pandas == filtered_transition_list