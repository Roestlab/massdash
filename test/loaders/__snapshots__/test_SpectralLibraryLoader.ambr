# serializer version: 1
# name: test_get_fragment_library_intensity[diann-DYASIDAAPEER-2-b3^1]
  0.2398542
# ---
# name: test_get_fragment_library_intensity[openswath-DYASIDAAPEER-2-y4/0.01y10^2/0.50]
  5399.8
# ---
# name: test_get_peptide_fragment_annotation_list[diann-DYASIDAAPEER-2]
  list([
    'b4^1',
    'y5^1',
    'b4^1',
    'y10^1',
    'b3^1',
    'b7^1',
    'y5^1',
    'b3^1',
    'y8^1',
    'b6^1',
    'y9^1',
    'b5^1',
    'y4^1',
    'b7^1',
    'y7^1',
    'y6^1',
    'y7^1',
    'y9^1',
    'b6^1',
    'y3^1',
    'y3^1',
    'b5^1',
    'b8^1',
    'y4^1',
    'y3^1',
    'y10^2',
    'y4^1',
  ])
# ---
# name: test_get_peptide_fragment_annotation_list[openswath-DYASIDAAPEER-2]
  list([
    'b3/0.01b7-36^2/-0.51',
    'y4/0.01y10^2/0.50',
    'y5/0.01y11-18^2/-0.99',
    'y7/-0.00',
    'y8/0.00',
    'y9/0.00b10-46/0.03',
  ])
# ---
# name: test_get_peptide_ion_mobility[diann-im-AFVDFLSDEIK-2]
  1.0276841808
# ---
# name: test_get_peptide_library_intensity[diann-DYASIDAAPEER-2]
  0.13932478
# ---
# name: test_get_peptide_library_intensity[openswath-DYASIDAAPEER-2]
  3475.6
# ---
# name: test_get_peptide_precursor_mz[diann-DYASIDAAPEER-2]
  668.80402
# ---
# name: test_get_peptide_precursor_mz[openswath-DYASIDAAPEER-2]
  668.804
# ---
# name: test_get_peptide_product_charge_list[diann-DYASIDAAPEER-2]
  list([
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    2,
    1,
  ])
# ---
# name: test_get_peptide_product_charge_list[openswath-DYASIDAAPEER-2]
  list([
    1,
    1,
    1,
    1,
    1,
    1,
  ])
# ---
# name: test_get_peptide_product_mz_list[diann-DYASIDAAPEER-2]
  list([
    437.16724,
    601.2951,
    419.15613,
    1058.5123,
    350.13519,
    736.31537,
    584.26855,
    332.12408,
    900.44324,
    665.27826,
    987.47528,
    550.25128,
    513.23145,
    718.30426,
    787.35919,
    672.33221,
    770.33264,
    969.46417,
    647.26715,
    433.20523,
    415.19412,
    532.24017,
    807.35248,
    512.24689,
    416.17868,
    529.76007,
    530.258,
  ])
# ---
# name: test_get_peptide_product_mz_list[openswath-DYASIDAAPEER-2]
  list([
    350.141,
    530.262,
    601.3,
    787.356,
    900.444,
    987.475,
  ])
# ---
# name: test_get_peptide_retention_time[openswath-DYASIDAAPEER-2]
  13.3
# ---
# name: test_get_unique_charge_states_per_peptide[diann]
  array([], dtype=int64)
# ---
# name: test_get_unique_charge_states_per_peptide[openswath]
  array([], dtype=int64)
# ---
# name: test_get_unique_peptides_per_protein[diann]
  array([], dtype=object)
# ---
# name: test_get_unique_peptides_per_protein[openswath]
  array([], dtype=object)
# ---
# name: test_get_unique_proteins[diann]
  array(['P0C0G7;P0DB18;P0DB19;P68777;Q5XDW3', 'P69952',
         'A2RFE3;B5XKM7;P0DA94;P0DA95;P69949;P69951;Q1J7I5;Q1JHQ6;Q48UF7'],
        dtype=object)
# ---
# name: test_get_unique_proteins[openswath]
  array(['Spyo_Exp3652_DDB_SeqID_514830', 'Spyo_Exp3652_DDB_SeqID_514851'],
        dtype=object)
# ---
# name: test_populateTransitionGroupFeatures[diann]
  NoneType(
  )
# ---
# name: test_populateTransitionGroupFeatures[openswath]
  NoneType(
  )
# ---
